### ISSUE TITLE
Feature request: Option to disable autoclosing tags behavior or provide a "strict" mode

### ISSUE THREAD

**Issue by Shoh** [zzzprojects/html-agility-pack#330]
### Description
I am working on a project that requires validating HTML and reporting on any errors it may find.  I was hoping to using `HtmlAgilityPack` to help with this but it seems it doesn't do exactly what I expect or was hoping for.  It looks like `HtmlAgilityPack` automatically closes any unclosed tags (when possible) and doesn't report unclosed tags as "errors" (at least when they're nested).  Unfortunately, this is preventing me from being able to use `HtmlAgilityPack` for this project because I actually need it to detect and report unclosed tags as errors instead of auto-fixing them.

Therefore, I am requesting that an option/feature be added to allow for turning off the "autoclosing tags" behavior or adding a "strict-mode" option that will report all errors instead of trying to auto-fix mistakes (such as unclosed tags).  If there is already a way to turn off the autoclosing tags behavior, I would love to hear about it!  Thank you!

### Fiddle or Project
Here is a fiddle demonstrating what I am seeing/expecting: https://dotnetfiddle.net/zvwy0n

### Further technical details
- HAP version: 1.11.12
- NET version (net472, netcore, etc.):  netcoreapp2.2

**Comment by JonathanMagnan** [comment]
Hello @Shoh ,

Thank you for reporting, we will look at it. I believe it will be possible to turn off the autoclosing with an option

Best Regards,

Jonathan

---

**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Entity Framework Classic](http://entityframework-classic.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/)

**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by EvgeniyZ** [comment]
@Shoh I believe you can easily identify not closed tags errors when you define html document like this:
`HtmlDocument document = new HtmlDocument
 {
                OptionAutoCloseOnEnd = true
 };`

`document.LoadHtml(html);`
`document.ParseErrors` will not be empty in case of html that have unclosed tags.

**Comment by Shoh** [comment]
Nice!  Thank you for the suggestion, @EvgeniyZ!  It looks like this should mostly work for my use cases so I'll be trying this out.

**Comment by EvgeniyZ** [comment]
@Shoh, you are welcome, so, please close this ticket if it works for you

**Comment by Shoh** [comment]
@EvgeniyZ I've tested it out and it looks to be working for my use cases but I'm curious what other side effects there are from specifying `OptionAutoCloseOnEnd`?  Does it change any other behavior?  I'm also wondering if you think a "strict-mode" would still be beneficial to have (i.e. all tags that require a closing tag must be closed or errors are reported) or do you think the `OptionAutoCloseOnEnd` already satisfies that?  Is this the actual purpose of `OptionAutoCloseOnEnd` (to catch unclosed tags errors)?

**Comment by EvgeniyZ** [comment]
@Shoh I don't know, I hope @JonathanMagnan can clarify that. I just faced similar problem, I wanted to catch errors if tags are not closed.
There's no doubt that `OptionAutoCloseOnEnd` changes html document, from docs:
`Defines if closing for non closed nodes must be done at the end or directly in the document.
Setting this to true can actually change how browsers render the page. Default is false.
`
@JonathanMagnan can you please describe what does mean - `can actually change how browsers rendeer the page.`

**Comment by JonathanMagnan** [comment]
Sorry for the long delay, we had to put all the free libraries on hold during September.

> can you please describe what does mean - can actually change how browsers rendeer the page.

It probably means that if you save the page as `HTML`, the render might be different from the page you originally loaded. 

> what other side effects there are from specifying OptionAutoCloseOnEnd

I'm not the original author of the library so besides looking at the source code and report what I found, I cannot really help you on this one. You might learn more about checking the source code on your side.

Basically, when we close a node, we usually close automatically all child nodes unless you turn off that option. So a fake "close" node is created when the child node is not closed `<div><p>I'm not closed!</div>`
